File: gateway-go/converter/schema.go
package converter

import (
	"fmt"
	"strings"
)

// CleanSchemaForGemini recursively cleans a JSON Schema for Gemini compatibility.
func CleanSchemaForGemini(schema map[string]any) map[string]any {
	if schema == nil {
		return nil
	}
	visited := make(map[string]bool)
	return cleanSchemaRecursive(schema, visited)
}

func cleanSchemaRecursive(schema map[string]any, visited map[string]bool) map[string]any {
	if schema == nil {
		return nil
	}

	// Detect circular references
	id := fmt.Sprintf("%p", schema)
	if visited[id] {
		return schema
	}
	visited[id] = true

	result := make(map[string]any)

	// Copy and transform fields
	for k, v := range schema {
		switch k {
		case "type":
			result["type"] = convertType(v)
		case "properties":
			if props, ok := v.(map[string]any); ok {
				cleaned := make(map[string]any)
				for pk, pv := range props {
					if pm, ok := pv.(map[string]any); ok {
						cleaned[pk] = cleanSchemaRecursive(pm, visited)
					} else {
						cleaned[pk] = pv
					}
				}
				result["properties"] = cleaned
			}
		case "items":
			if items, ok := v.(map[string]any); ok {
				result["items"] = cleanSchemaRecursive(items, visited)
			}
		case "allOf":
			// Merge all schemas in allOf
			if arr, ok := v.([]any); ok {
				merged := mergeAllOf(arr, visited)
				for mk, mv := range merged {
					result[mk] = mv
				}
			}
		case "anyOf":
			// Convert anyOf to enum if all items have const
			if arr, ok := v.([]any); ok {
				if enums := extractEnumFromAnyOf(arr); enums != nil {
					result["enum"] = enums
				}
			}
		case "default":
			// Move default to description
			if desc, ok := result["description"].(string); ok {
				result["description"] = fmt.Sprintf("%s (Default: %v)", desc, v)
			} else {
				result["description"] = fmt.Sprintf("(Default: %v)", v)
			}
		case "required", "description", "enum", "format", "nullable":
			result[k] = v
		case "$defs", "definitions", "$schema", "$id", "const", "oneOf", "strict":
			// Remove unsupported fields
		default:
			result[k] = v
		}
	}

	return result
}

func convertType(v any) string {
	switch t := v.(type) {
	case string:
		return typeToGemini(t)
	case []any:
		// Extract non-null type from type array
		for _, item := range t {
			if s, ok := item.(string); ok && s != "null" {
				return typeToGemini(s)
			}
		}
		return "STRING"
	default:
		return "STRING"
	}
}

func typeToGemini(t string) string {
	switch strings.ToLower(t) {
	case "string":
		return "STRING"
	case "number":
		return "NUMBER"
	case "integer":
		return "INTEGER"
	case "boolean":
		return "BOOLEAN"
	case "array":
		return "ARRAY"
	case "object":
		return "OBJECT"
	default:
		return "STRING"
	}
}

func mergeAllOf(arr []any, visited map[string]bool) map[string]any {
	merged := make(map[string]any)
	mergedProps := make(map[string]any)
	var mergedRequired []any

	for _, item := range arr {
		if m, ok := item.(map[string]any); ok {
			cleaned := cleanSchemaRecursive(m, visited)
			for k, v := range cleaned {
				switch k {
				case "properties":
					if props, ok := v.(map[string]any); ok {
						for pk, pv := range props {
							mergedProps[pk] = pv
						}
					}
				case "required":
					if req, ok := v.([]any); ok {
						mergedRequired = append(mergedRequired, req...)
					}
				default:
					merged[k] = v
				}
			}
		}
	}

	if len(mergedProps) > 0 {
		merged["properties"] = mergedProps
	}
	if len(mergedRequired) > 0 {
		merged["required"] = mergedRequired
	}

	return merged
}

func extractEnumFromAnyOf(arr []any) []any {
	var enums []any
	for _, item := range arr {
		if m, ok := item.(map[string]any); ok {
			if c, ok := m["const"]; ok {
				enums = append(enums, c)
			} else {
				return nil // Not all items have const
			}
		}
	}
	return enums
}

File: gateway-node/src/converter/schema.ts
export function cleanSchemaForGemini(schema: Record<string, any> | undefined): Record<string, any> | undefined {
  if (!schema) return undefined;
  const visited = new Set<any>();
  return cleanRecursive(schema, visited);
}

function cleanRecursive(schema: Record<string, any>, visited: Set<any>): Record<string, any> {
  if (visited.has(schema)) return schema;
  visited.add(schema);

  const result: Record<string, any> = {};

  for (const [k, v] of Object.entries(schema)) {
    switch (k) {
      case 'type':
        result.type = convertType(v);
        break;
      case 'properties':
        if (typeof v === 'object' && v !== null) {
          const cleaned: Record<string, any> = {};
          for (const [pk, pv] of Object.entries(v)) {
            cleaned[pk] = typeof pv === 'object' && pv !== null ? cleanRecursive(pv as Record<string, any>, visited) : pv;
          }
          result.properties = cleaned;
        }
        break;
      case 'items':
        if (typeof v === 'object' && v !== null) {
          result.items = cleanRecursive(v as Record<string, any>, visited);
        }
        break;
      case 'allOf':
        if (Array.isArray(v)) {
          const merged = mergeAllOf(v, visited);
          Object.assign(result, merged);
        }
        break;
      case 'anyOf':
        if (Array.isArray(v)) {
          const enums = extractEnumFromAnyOf(v);
          if (enums) result.enum = enums;
        }
        break;
      case 'default':
        result.description = result.description
          ? `${result.description} (Default: ${v})`
          : `(Default: ${v})`;
        break;
      case 'required':
      case 'description':
      case 'enum':
      case 'format':
      case 'nullable':
        result[k] = v;
        break;
      case '$defs':
      case 'definitions':
      case '$schema':
      case '$id':
      case 'const':
      case 'oneOf':
      case 'strict':
        break;
      default:
        result[k] = v;
    }
  }
  return result;
}

function convertType(v: any): string {
  if (typeof v === 'string') return typeToGemini(v);
  if (Array.isArray(v)) {
    const nonNull = v.find((t: string) => t !== 'null');
    return nonNull ? typeToGemini(nonNull) : 'STRING';
  }
  return 'STRING';
}

function typeToGemini(t: string): string {
  const map: Record<string, string> = {
    string: 'STRING', number: 'NUMBER', integer: 'INTEGER',
    boolean: 'BOOLEAN', array: 'ARRAY', object: 'OBJECT',
  };
  return map[t.toLowerCase()] || 'STRING';
}

function mergeAllOf(arr: any[], visited: Set<any>): Record<string, any> {
  const merged: Record<string, any> = {};
  const mergedProps: Record<string, any> = {};
  const mergedRequired: any[] = [];

  for (const item of arr) {
    if (typeof item === 'object' && item !== null) {
      const cleaned = cleanRecursive(item, visited);
      for (const [k, v] of Object.entries(cleaned)) {
        if (k === 'properties' && typeof v === 'object') {
          Object.assign(mergedProps, v);
        } else if (k === 'required' && Array.isArray(v)) {
          mergedRequired.push(...v);
        } else {
          merged[k] = v;
        }
      }
    }
  }
  if (Object.keys(mergedProps).length > 0) merged.properties = mergedProps;
  if (mergedRequired.length > 0) merged.required = mergedRequired;
  return merged;
}

function extractEnumFromAnyOf(arr: any[]): any[] | null {
  const enums: any[] = [];
  for (const item of arr) {
    if (typeof item === 'object' && item !== null && 'const' in item) {
      enums.push(item.const);
    } else {
      return null;
    }
  }
  return enums;
}
